logstash-core/src/main/java/org/logstash/ext/JrubyMemoryReadClientExt.java

```diff
@@ -0,0 +1,179 @@
+        return this;
+    }
+
+    @JRubyMethod(name = "set_pipeline_metric", required = 1)
+    public IRubyObject setPipelineMetric(final ThreadContext context, IRubyObject metric) {
+        pipelineMetricOut = getCounter(metric, OUT_KEY);
+        pipelineMetricFiltered = getCounter(metric, FILTERED_KEY);
+        pipelineMetricTime = getCounter(metric, DURATION_IN_MILLIS_KEY);
+        return this;
+    }
+
+    @JRubyMethod(name = "inflight_batches")
+    public IRubyObject rubyGetInflightBatches(final ThreadContext context) {
+        return RubyHash.newHash(context.runtime, inflightBatches, RubyBasicObject.UNDEF);
+    }
+
+    // create a new, empty batch
+    @JRubyMethod(name = "new_batch")
+    public IRubyObject newBatch(final ThreadContext context) {
+        return JrubyMemoryReadBatchExt.create();
+    }
+
+    @JRubyMethod(name = "read_batch")
+    public IRubyObject readBatch(final ThreadContext context) throws InterruptedException {
+        JrubyMemoryReadBatchExt batch = JrubyMemoryReadBatchExt.create(
+                LsQueueUtils.drain(queue, batchSize, waitForNanos));
+        startMetrics(batch);
+        return batch;
+    }
+
+    @JRubyMethod(name = "start_metrics")
+    public IRubyObject rubyStartMetrics(final ThreadContext context, IRubyObject batch) {
+        startMetrics((JrubyMemoryReadBatchExt) batch);
+        return this;
+    }
+
+    private void startMetrics(JrubyMemoryReadBatchExt batch) {
+        inflightBatches.put(Thread.currentThread().getId(), batch);
+        inflightClocks.put(Thread.currentThread().getId(), System.nanoTime());
+    }
+
+    @JRubyMethod(name = "close_batch", required = 1)
+    public IRubyObject closeBatch(final ThreadContext context, IRubyObject batch) {
+        JrubyMemoryReadBatchExt typedBatch = (JrubyMemoryReadBatchExt) batch;
+        inflightBatches.remove(Thread.currentThread().getId());
+        Long startTime = inflightClocks.remove(Thread.currentThread().getId());
+        if (startTime != null && typedBatch.filteredSize() > 0) {
+            // stop timer and record metrics iff the batch is non-empty.
+            long elapsedTimeMillis = (System.nanoTime() - startTime) / 1_000_000;
+            eventMetricTime.increment(elapsedTimeMillis);
+            pipelineMetricTime.increment(elapsedTimeMillis);
+        }
+        return this;
+    }
+
+    @JRubyMethod(name = "add_filtered_metrics", required = 1)
+    public IRubyObject addFilteredMetrics(final ThreadContext context, IRubyObject filteredSize) {
+        eventMetricFiltered.increment(((RubyNumeric) filteredSize).getIntValue());
+        pipelineMetricFiltered.increment(((RubyNumeric) filteredSize).getIntValue());
+        return this;
+    }
+
+    @JRubyMethod(name = "add_output_metrics", required = 1)
+    public IRubyObject addOutputMetrics(final ThreadContext context, IRubyObject filteredSize) {
+        eventMetricOut.increment(((RubyNumeric) filteredSize).getIntValue());
```

Comment:
@danhermann 

small improvement if you have a sec :) Just cache:

```java
(RubyNumeric) filteredSize).getIntValue()
```

to a variable :)